Sources/MongoSwift/ConnectionPool.swift

```diff
@@ -0,0 +1,115 @@
+import mongoc
+
+/// A connection to the database.
+internal struct Connection {
+    /// Pointer to the underlying `mongoc_client_t`.
+    internal let clientHandle: OpaquePointer
+
+    internal init(_ clientHandle: OpaquePointer) {
+        self.clientHandle = clientHandle
+    }
+}
+
+/// A pool of one or more connections.
+internal class ConnectionPool {
+    /// Represents the mode of a `ConnectionPool`.
+    private enum Mode {
```

Comment:
I'm fine with moving `closed` out and using some boolean flag to indicate that it's closed. it's kind of funky that we will have to keep around the pointers to already-destroyed mongoc clients / pools that the modes store, but as long as we don't try to use them it seems fine to do.

re: closing the pool, see my other comment - you are correct that we never enter this state right now. 
however, one of the [CMAP requirements](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#connection-pool) for the pool is:
> A Pool MUST be able to be manually closed.

I'm not sure why that is though, do you know?

that said, we could just get rid of the closed mode/manual close behavior altogether and worry about that particular CMAP requirement when we actually make this a true CMAP pool in the future.

Reply:
>but as long as we don't try to use them it seems fine to do.

Having a `closed` case enforces this, which is why I kind of like including it. I do admit that it really isn't a "mode" though, so maybe it makes more sense to rename `Mode` to `State` or `Status`?

Reply:
we could also make the mode an optional, and then set it to nil when we close the pool. no boolean needed. though then we need to unwrap it any time we want to use it

Reply:
I like the idea of an optional, but now that I'm toying with the idea it doesn't seem to buy us much. While I think it reads somewhat better, you still need to reset the optional meaning we can't make the mode immutable:

```
internal func close() {
    guard let mode = self.mode else {
         return
    }

    switch mode {
    case let .single(clientHandle):
        mongoc_client_destroy(clientHandle)
    case .pooled:
        fatalError("Pooled mode is not yet implemented")
    }

    self.mode = nil
}
```

I don't feel _too_ strongly about this. Patrick's idea sounds fine, or maybe we can just rename `closed` to something that sounds less like a state? 

I'm not sure why CMAP requires manual closing of the pool, but it seems like at very least it will be useful for testing (I'm sure that spec uses this feature quite a lot probably).

Reply:
I'm fine with renaming `Mode`->`State`. also fine with renaming the `closed` case though I don't immediately have an idea for a better name.

yes, there are many tests involving behavior of a closed pool in CMAP. I don't think we will ever actually exercise the manual close feature ourselves (at least not while `MongoClient`s own pools) but I don't see harm in making it possible to manually close the pool.

Reply:
renamed `closed` -> `none`